### DISCUSSION TITLE
Is there a MicroPython equivalent of the 'rpi_ws281x' library for (ESP32) microcontrollers? (Not NeoPixel)

### DISCUSSION THREAD

**Discussion by ion-mironov** [micropython/micropython#13783] [Libraries & Drivers] [answered]
Hello, everyone! I've been having a tough time finding an answer/solution to my current problem: In the past, my project consisted of running three LED panels (256 LEDs each) off of my Raspberry Pi 4B with the **rpi_ws281x library**. After learning that the more LEDs I try to control the slower their animations became, I got myself my first ESP32 microcontroller (ESP32-WROOM-32D, to be exact.) to control the LEDs via that. However, from what I've found so far, there does not seem to be an **rpi_ws281x** equivalent library in MicroPython for microcontrollers. The only thing I've found is NeoPixel, but I'm not a huge fan of that library, mainly because I cannot find any documentation that covers all of its attributes and properties (so that I can "translate" my **rpi_ws281x** code to NeoPixel.)

Is there a MicroPython version of **rpi_ws281x**? I've only found a couple but their code looks to be completely convoluted and confusing.

**Comment by scruss** [comment, marked as answer]
No, there isn't an exact port. rpi_ws281x is hardware specific to the Raspberry Pi. You're going to have to do some porting of your code, especially considering memory limitations of the ESP32 compared to the Raspberry Pi

The docs for the standard MicroPython NeoPixel driver:

* [Quick reference for the ESP32 — NeoPixel](https://docs.micropython.org/en/v1.22.0/esp32/quickref.html#neopixel-and-apa106-driver)
* [micropython/docs/library/neopixel.rst - Constructors](https://github.com/micropython/micropython/blob/678707c8b07323c5b914778708a2858387c3b60c/docs/library/neopixel.rst#constructors)

The one feature that's consistently missing from MicroPython NeoPixel is setting brightness/intensity. MicroPython defaults to 100% brightness. The (possibly rather old) [Warringer/micropython-rgbled](https://github.com/Warringer/micropython-rgbled) wrapper seems to address this, but not in a way that's compatible with the rpi_ws281x library

**Reply by ion-mironov** [reply]
Dammit. Seems there's always a Catch-22 when it comes to stuff like this; control the LEDs via the Pi and you'll get slower animations (with lots of LEDs). Control the LEDs via the microcontroller and you'll get faster animations but have memory limitations.

At least I was able to conjure up a brightness code that would allow me to set the brightness for individual LED animations:

```
import time
import machine
import neopixel

""" Global LED panel configuration """
LED_PIN = machine.Pin(18, machine.Pin.OUT)
LED_COUNT = 768

amber = (255, 100, 0)
red = (255, 0, 0)
green = (0, 255, 0)
blue = (0, 0, 255)
off = (0, 0, 0)

""" Initialize the LED pixels """
pixels = neopixel.NeoPixel(LED_PIN, LED_COUNT)

def setPixelColor(color, brightness = 1.0):
	r, g, b = color
	r = int(r * brightness)
	g = int(g * brightness)
	b = int(b * brightness)
	for i in range(LED_COUNT):
		pixels[i] = (r, g, b)
	pixels.write()

setPixelColor(amber, 0.1)  
time.sleep(1)
setPixelColor(blue, 1.0)
time.sleep(1)
setPixelColor(off)
```

**Reply by scruss** [reply]
you can do `pixels.fill(...)` to set the whole strip to the same colour (if the second doc I linked to is to be believed)

**Reply by ion-mironov** [reply]
Good catch, I forgot about that. I had it in a newer version of that code I posted, but forgot to include it.
